docs/examples/hypervisors/kvm/modules/kvm-hypervisor/main.tf

```diff
@@ -0,0 +1,169 @@
+  }
+
+  provisioner "remote-exec" {
+    inline = [
+      "chmod +x ~/install-kvm-dependencies.sh",
+      "sudo sh -x ~/install-kvm-dependencies.sh",
+    ]
+  }
+}
+
+resource "null_resource" "restart-instance" {
+  depends_on = ["null_resource.install-kvm-dependencies"]
+
+  connection {
+    type        = "ssh"
+    user        = "opc"
+    private_key = "${var.private_key}"
+    host        = "${var.host}"
+    timeout     = "30m"
+  }
+
+  provisioner "remote-exec" {
+    inline = [
+      "echo waiting...",
+      "echo restarting instance",
+      "sudo shutdown -r now",
+    ]
+  }
+}
+
+resource "null_resource" "waiting_for_reboot" {
+  depends_on = ["null_resource.restart-instance"]
+
+  connection {
+    type        = "ssh"
+    user        = "opc"
+    private_key = "${var.private_key}"
+    host        = "${var.host}"
+    timeout     = "30m"
+  }
+
+  provisioner "remote-exec" {
+    inline = [
+      "ls -la ~/",
+    ]
+  }
+}
+
+resource "null_resource" "configure-secondary-vnics" {
+  depends_on = ["null_resource.waiting_for_reboot"]
+
+  connection {
+    type        = "ssh"
+    user        = "opc"
+    private_key = "${var.private_key}"
+    host        = "${var.host}"
+    timeout     = "30m"
+  }
+
+  provisioner "file" {
+    source      = "${path.module}/scripts/configure-secondary-vnics.sh"
+    destination = "~/configure-secondary-vnics.sh"
+  }
+
+  provisioner "file" {
+    source      = "${path.module}/scripts/secondary_vnics.service"
+    destination = "~/secondary_vnics.service"
+  }
+
+  provisioner "remote-exec" {
+    inline = [
+      "sudo mv ~/configure-secondary-vnics.sh /usr/bin/configure-secondary-vnics.sh",
+      "sudo mv ~/secondary_vnics.service /etc/systemd/system/secondary_vnics.service",
+      "chmod 744 /usr/bin/configure-secondary-vnics.sh",
+      "sudo systemctl daemon-reload",
+      "sudo systemctl enable secondary_vnics.service",
+      "sudo systemctl start secondary_vnics.service",
+      "sudo ifconfig",
+    ]
+  }
+}
+
+resource "null_resource" "upload-kvm-qcow2-image" {
+  depends_on = ["null_resource.configure-secondary-vnics"]
+
+  connection {
+    type        = "ssh"
+    user        = "opc"
+    private_key = "${var.private_key}"
+    host        = "${var.host}"
+    timeout     = "30m"
+  }
+
+  provisioner "remote-exec" {
+    inline = [
+      "wget ${var.qcow2_image_url}",
+      "sudo mkdir ${var.qcow2_image_target_path}",
+      "sudo mv ~/${var.qcow2_image_filename} ${var.qcow2_image_target_path}${var.qcow2_image_filename}",
+    ]
+  }
+}
+
+resource "null_resource" "create-kvm-domain" {
+  depends_on = ["null_resource.upload-kvm-qcow2-image"]
+
+  connection {
+    type        = "ssh"
+    user        = "opc"
+    private_key = "${var.private_key}"
+    host        = "${var.host}"
+    timeout     = "30m"
+  }
+
+  provisioner "remote-exec" {
+    inline = [
+      "${data.template_file.generate_virsh_attach.rendered}",
```

Comment:
How does this get executed for all 3 VNICs for the guest VM?

Reply:
This is a template file. Terraform will output the content of [generate_virsh_attach.sh.tpl](./generate_virsh_attach.sh.tpl) in the host instance at runtime. The template file creates a bash script which will retrieve all the vnic data from the metadata repository for every vnic registered.